src/transformers/retrieval_rag.py

```diff
@@ -0,0 +1,535 @@
+            )
+        self._index_initialize = True
+
+    def get_doc_dicts(self, doc_ids):
+        return [self.index[doc_ids[i].tolist()] for i in range(doc_ids.shape[0])]
+
+    def get_top_docs(self, question_hidden_states, n_docs=5):
+        _, docs = self.index.get_nearest_examples_batch("embeddings", question_hidden_states, n_docs)
+        ids = [[int(i) for i in doc["id"]] for doc in docs]
+        vectors = [doc["embeddings"] for doc in docs]
+        return torch.tensor(ids), torch.tensor(vectors)
+
+
+class RagRetriever(object):
+    """
+    A distributed retriever built on top of the ``torch.distributed`` communication package. During training all workers
+    initalize their own instance of the retriever, however, only the main worker loads the index into memory. The index is stored
+    in cpu memory. The index will also work well in a non-distributed setup.
+
+    Args:
+        config (:class:`~transformers.RagConfig`):
+            The configuration of the RAG model this Retriever is used with. Contains parameters indicating which ``Index`` to build.
+    """
+
+    def __init__(self, config, question_encoder_tokenizer=None, generator_tokenizer=None):
+        super().__init__()
+        self.retriever = (
+            LegacyIndex(
+                config.retrieval_vector_size,
+                config.index_path or LEGACY_INDEX_PATH,
+            )
+            if config.index_name == "legacy"
+            else HFIndex(
+                config.dataset, config.dataset_split, config.index_name, config.index_path, config.use_dummy_dataset
+            )
+        )
+        # TODO(quentin) use RagTokenizer once the API is defined
+        self.generator_tokenizer = generator_tokenizer
+        self.question_encoder_tokenizer = question_encoder_tokenizer
+
+        self.process_group = None
+        self.n_docs = config.n_docs
+        self.batch_size = config.retrieval_batch_size
+
+        if torch.cuda.is_available():
+            self.batch_size *= torch.cuda.device_count()
+
+        self.config = config
+
+    @classmethod
+    def from_pretrained(cls, retriever_name_or_path, **kwargs):
+        config = RagConfig.from_pretrained(retriever_name_or_path, **kwargs)
+        rag_tokenizer = RagTokenizer.from_pretrained(retriever_name_or_path, config=config)
+        question_encoder_tokenizer = rag_tokenizer.question_encoder
+        generator_tokenizer = rag_tokenizer.generator
+        return cls(
+            config, question_encoder_tokenizer=question_encoder_tokenizer, generator_tokenizer=generator_tokenizer
+        )
+
+    def save_pretrained(self, save_directory):
+        self.config.save_pretrained(save_directory)
+        rag_tokenizer = RagTokenizer(
+            question_encoder_tokenizer=self.question_encoder_tokenizer,
+            generator_tokenizer=self.generator_tokenizer,
+        )
+        rag_tokenizer.save_pretrained(save_directory)
+
+    def init_retrieval(self, distributed_port):
```

Comment:
I'm wondering if we should not have this distributed function in the examples rather than the core of the lib.
This way it will be easier to use other distributed frameworks like Ray and also to use the retriever with Tensorflow.

Can be done in a follow-up as well maybe.